### PR TITLE
Add Tealite: pure-Python TransformerEngine for ROCm/AMD GPUs

### PR DESCRIPTION
**Pull request by jayfurmanek**
# Description

Introduces **`tealite`** — a pure-Python, drop-in replacement for the
`transformer_engine_torch` C++ extension, targeting **ROCm / AMD GPUs** and
PyTorch only. Activated by setting `NVTE_LITE=1`; when unset, none of the
lite code is imported and full TE behavior is unchanged.

**Why:** the full build compiles hundreds of C++ / CUDA / HIP sources via
CMake, couples tightly to toolchain versions, and slows down rapid kernel
iteration on ROCm. Tealite sidesteps all of that by dispatching to
[AITER](https://github.com/ROCm/aiter) kernels (CK + Triton, installed via
`pip install amd-aiter` — no submodule), standalone Triton kernels,
`torch._scaled_mm` (hipBLASLt-backed) for FP8 GEMMs, and PyTorch-native
fallbacks. Build a wheel in seconds, not minutes.

Note I will squash upon review as needed. 

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation change

## Changes

**New `_lite/` package (27 files, ~14k LOC) at
`transformer_engine/pytorch/_lite/`:**

- **GEMM** (`gemm.py`, `grouped_gemm.py`): pluggable backend dispatcher
  (`NVTE_LITE_GEMM_BACKEND` = `pytorch` / `triton` / `ck`); FP8 routes
  through `torch._scaled_mm` with AITER fallback for shapes hipBLASLt can't
  serve (per-row reduction-axis, K-not-div-16, unsupported dtype combos).
- **Norms** (`norms.py`): RMSNorm/LayerNorm via AITER Triton with TE-Triton
  and PyTorch fallbacks. Per-row dynamic-quant fusion
  (`rmsnorm2d_fwd_with_dynamicquant`) is a lite-only CurrentScaling
  optimization not available in the full build.
- **Quantization** (`quantize.py`): FP8 / MXFP8 / MXFP4 cast, transpose,
  block-scaling. Fused amax/scale update via a single Triton
  multi-tensor-apply kernel mirroring `delayed_scaling.cu::kernel_bulk`
  (replaces ~530k tiny elementwise launches/iter).
- **Attention** (`attention.py`): SDPA, AITER CK (`fmha_v3_fwd`/`fmha_v3_bwd`
  AOT), and flash-attn package backends.
- **Activations** (`activations.py`): AITER fused gated activations
  (swiglu, geglu, reglu, qgeglu) with PyTorch fallbacks.
- **RoPE** (`rope.py`): AITER Triton kernels with CP support.
- **Compound modules** (`fused_layernorm_linear.py`,
  `fused_layernorm_mlp.py`): pure-Python `torch.autograd.Function`
  subclasses with full DelayedScaling / CurrentScaling / MXFP8 support and
  FSDP2 integration via the inherited `FSDPAGTensor` wrap.
- **Distributed**: `comm.py`, `mori_ep.py` (MORI-based expert parallelism
  for MoE), `context_parallel.py` (THD/BSHD helpers; CP is wired in
  attention + RoPE).
- **Other**: dropout, softmax, padding, transpose, multi-tensor optimizer
  ops, router (fused TopK + softmax), permutation, MoE grouped GEMM.

**Minimal full-build touchpoints (5 files, ~265 LOC) — all env-gated or
inert when `NVTE_LITE` is unset:**

| File | Lines | What |
|---|---:|---|
| `module/__init__.py` | +7 | env-gated swap of `LayerNormLinear` / `LayerNormMLP` for lite versions when `NVTE_LITE=1` |
| `module/base.py` | +66/-13 | `FSDPAGTensor` weight-wrap plumbing (gated on `IS_HIP_EXTENSION`) |
| `quantization.py` | +2 | `_contig_diag.tick_step()` hook for the optional materialize-attribution harness (no-op when env unset) |
| `triton/fused_router.py` | +183 (new) | fused TopK + softmax router |
| `triton_kernels/grouped_gemm.py` | +9/-4 | grouped GEMM helper tweaks |

**Documentation**: in-tree `_lite/README.md` (full feature matrix, env
vars, status notes per subsystem, known gaps) and `_lite/SKILLS.md`
(operational tips).

## Performance

End-to-end FP8 training on **8×MI300X / LLaMA-3-8B / seq=2048** at the
same TE commit (2026-05-01):

| Mode | iter | tok/GPU/s |
|---|---:|---:|
| full | 1712 ms | 9567 |
| lite | 1712 ms | 9569 |

At parity. Lite uses the same hipBLASLt FP8 kernels via `torch._scaled_mm`
for fwd + dgrad, AITER Triton for wgrad shapes hipBLASLt can't serve, and
the fused Triton amax/scale update.

## Known limitations (documented in `_lite/README.md`)

- **No tensor parallelism / Megatron-style sequence parallelism** — the
  compound modules accept `tp_size`/`tp_group`/`parallel_mode`/
  `sequence_parallel` for API compat but hardcode `tp_size=1`. The
  multi-node story for tealite is FSDP/HSDP-shaped, not TP-shaped.
- **HSDP (2D device mesh) not plumbed** — FSDP2 with a 1D mesh works
  and is tested.
- **MoE FP8 path** — BF16 grouped works via the `tex` hot-swap into
  `_lite/grouped_gemm.py`; FP8 grouped is blocked on a Triton GMM dtype
  mismatch and `aiter.fused_moe` wiring.
- **No comm-overlap / userbuffers** — stubs raise.
- **No cuDNN backend, no pre-tuned hidden sizes** — auto-tune only.
- **Per-row CurrentScaling LayerNorm** — RMSNorm only (AITER doesn't ship
  a `layernorm2d_fwd_with_dynamicquant`).

## Activation

```bash
pip install amd-aiter
NVTE_LITE=1 python your_training_script.py
```

# Checklist:

- [x] The functionality is complete (for the documented scope)
- [x] I have made corresponding changes to the documentation
      (`_lite/README.md`, `_lite/SKILLS.md`)
- [x] I have added tests that prove my feature works
      (`tests/pytorch/test_lite.py`, 4967 lines — GEMM backend matrix,
      FP8 dispatch, per-row CurrentScaling end-to-end, MoE smoke,
      FSDP2 wrap verification, attention backends, norms + quant,
      activations, multi-tensor ops)
- [x] New and existing unit tests pass locally with my changes
- [x] My changes generate no new warnings (when `NVTE_LITE=1`)
- [ ] I have read and followed the contributing guidelines

### CODE REVIEW

File: build_tools/utils.py [hunk 1]
```diff
@@ -415,13 +415,14 @@ def get_frameworks() -> List[str]:
         if framework not in supported_frameworks:
             raise ValueError(f"Transformer Engine does not support framework={framework}")
 
-    if rocm_build():
+    if rocm_build() and not bool(int(os.getenv("NVTE_LITE_ONLY", "0"))):
         _unsupported_frameworks = []
         if "pytorch" in _frameworks:
             try:
-                from torch.utils.cpp_extension import IS_HIP_EXTENSION
+                import torch.utils.cpp_extension
```
Comment (ipanfilo) [inline build_tools/utils.py:422]: It does not seem changing behavior - ImportError should be raised if IS_HIP_EXTENSION is not defined in cpp_extension

File: tests/pytorch/attention/run_lite_cp_test.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
+# Copyright (c) 2026, Advanced Micro Devices, Inc. All rights reserved.
+# Copyright (c) 2022-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (ipanfilo) [inline tests/pytorch/attention/run_lite_cp_test.py:2]: If it is AMD only file then AMD only copyright should be here with MIT license

File: tests/pytorch/attention/test_lite_cp.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
+# Copyright (c) 2026, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (ipanfilo) [inline tests/pytorch/attention/test_lite_cp.py:1]: The same as in run_lite_cp_test.py

File: tests/pytorch/test_fused_rope.py
Comment (ipanfilo) [inline tests/pytorch/test_fused_rope.py:1]: AMD copyright is needed

File: transformer_engine/common/__init__.py [hunk 4]
```diff
@@ -402,6 +411,9 @@ def _load_cuda_library(lib_name: str):
 
 @functools.cache
 def is_fp8_fnuz():
+    if _TE_LIB_CTYPES is None:
+        # Lite mode: assume FNUZ based on ROCm convention
```
Comment (ipanfilo) [inline transformer_engine/common/__init__.py:415]: FNUZ is MI300 only
